client.hpp

```diff
@@ -13,11 +18,40 @@ class Client {
   Client(int socket_desc);
 public:
   Client(const std::string& address, const std::string& port);
-  void write(const char* data, const size_t& length);
   Client& operator<<(const std::string& data);
   Client& operator>>(std::string& result);
-  std::vector<char> read();
-  std::string read_str();
+  std::string read_str() const;
+
+  template<typename T>
+  std::vector<T> read() const {
+    int len = 0;
+    ioctl(*m_socket_ptr, FIONREAD, &len);
+    if (len == 0) {
+      return std::vector<T>();
+    }
+    std::vector<T> result(len);
```

Comment:
receiving 3 `long`s causes `len = 3 * sizeof(long) = 24` and therefore vector is always `sizeof(T)` times longer using unnecessary memory space.